### ISSUE TITLE
Machine Learning Project Report: Customer Classification

### ISSUE THREAD

**Issue by shboulmohmmad353-droid** [shboulmohmmad353-droid/Ticket-481#1]
Machine Learning Project Report: Customer Classification

Project Title:

Customer Classification Based on Shopping Patterns

Problem Statement:

This project focuses on classifying customers based on their shopping patterns using machine learning. By analyzing customer behavior, the goal is to create a model that predicts customer categories such as their likelihood of subscription renewal or their preferred shopping frequency. This classification can help businesses target customers more effectively and personalize their marketing strategies.

Dataset Description:

Dataset Name: Shopping Trends Dataset (provided as shopping_trends.csv).

Dataset Size: 3900 rows, 19 columns.

Features:

Demographic Information: Age, Gender, Location.

Purchase Details: Item Purchased, Category, Purchase Amount (USD), Size, Color, Season, Review Rating.

Behavioral Information: Subscription Status, Payment Method, Previous Purchases, Frequency of Purchases.

Target Variables:

Primary Target: Subscription Status (Yes/No).

Secondary Target: Frequency of Purchases (Weekly, Monthly, Annually, etc.).

Preprocessing Steps:

Handle categorical variables via encoding (e.g., one-hot encoding for Gender, Location).

Normalize numerical variables such as Age and Purchase Amount.

Handle missing data (if any).

Methods:

Three machine learning models will be implemented for comparison:

Decision Tree Classifier:

Chosen for its interpretability and effectiveness in handling mixed data types.

Support Vector Machine (SVM):

A robust model for classification tasks with non-linear decision boundaries.

Logistic Regression:

A baseline model for binary classification problems with straightforward implementation.

The implementation will include the following:

Data preprocessing (normalization, encoding, and splitting into training and testing sets).

Model training and evaluation.

Hyperparameter tuning for optimal performance.

Experimental Design:

Datasets:

Training Set: 70% of the dataset.

Validation Set: 15% of the dataset.

Test Set: 15% of the dataset.

Baselines:

Use Logistic Regression as a baseline for comparison with more complex models.

Metrics:

Accuracy: Percentage of correct predictions.

Precision: Ability to correctly predict positive instances.

Recall: Ability to capture all positive instances.

F1-Score: Harmonic mean of Precision and Recall.

Experimental Results:

Placeholder Results:

Decision Tree: Expected accuracy ~85% based on initial data patterns.

SVM: Expected accuracy ~90% with proper feature scaling.

Logistic Regression: Expected accuracy ~80% as a baseline.

Graphs and tables will be included in the final report to visualize performance comparisons.

Conclusions:

Preliminary Findings:

Subscription status is influenced by demographic and behavioral features.

Models are expected to perform well given the structured nature of the dataset.

Anticipated Challenges:

Balancing imbalanced classes if the subscription status is skewed.

Dealing with categorical features with high cardinality (e.g., Location).

Next Steps:

Complete data preprocessing.

Implement and evaluate models.

Analyze results and refine methods based on performance.

Appendices:

Code files and processed datasets will be included in the final submission.

Due Date: January 8th, 2025

**Comment by shboulmohmmad353-droid** [comment]
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, accuracy_score, confusion_matrix
import matplotlib.pyplot as plt
import seaborn as sns

data = pd.read_csv("")
data = pd.read_csv("")

X = data.drop(columns=["Subscription Status"])
y = data["Subscription Status"].map({"Yes": 1, "No": 0})  

categorical_columns = ["Gender", "Location", "Item Purchased", "Category", "Size", "Color", "Season", 
                       "Payment Method", "Shipping Type", "Discount Applied", "Promo Code Used", 
                       "Preferred Payment Method", "Frequency of Purchases"]
numerical_columns = ["Age", "Purchase Amount (USD)", "Review Rating", "Previous Purchases"]

categorical_transformer = OneHotEncoder(handle_unknown='ignore')
numerical_transformer = StandardScaler()

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numerical_transformer, numerical_columns),
        ('cat', categorical_transformer, categorical_columns)
    ]
)

models = {
    "Decision Tree": DecisionTreeClassifier(random_state=42),
    "SVM": SVC(probability=True, random_state=42),
    "Logistic Regression": LogisticRegression(max_iter=1000, random_state=42)
}

accuracies = []
for name, model in models.items():
    pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                               ('classifier', model)])
    
    pipeline.fit(X_train, y_train)
    y_pred = pipeline.predict(X_test)
    
    accuracy = accuracy_score(y_test, y_pred)
    accuracies.append(accuracy)
    
    print(f"Model: {name}")
    print("Accuracy:", accuracy)
    print("Classification Report:\n", classification_report(y_test, y_pred))
    
    cm = confusion_matrix(y_test, y_pred)
    plt.figure(figsize=(8, 6))
    sns.heatmap(cm, annot=True, fmt='d', cmap='Blues', xticklabels=["No", "Yes"], yticklabels=["No", "Yes"])
    plt.xlabel('Predicted')
    plt.ylabel('Actual')
    plt.title(f'Confusion Matrix for {name}')
    plt.show()
    
    print("-" * 50)

model_names = list(models.keys())
plt.bar(model_names, accuracies, color=['blue', 'green', 'orange'])
plt.xlabel('Models')
plt.ylabel('Accuracy')
plt.title('Model Comparison')
plt.ylim(0, 1)  
plt.show()
